### PR TITLE
Logging for snare (#73)

### LINKED ISSUE #73: Log timestamp and source IP addresses
The log should at least include timestamp, source IP address and the URI.

### PR DESCRIPTION
**Pull request by mzfr**
Fixes #73 

I added a logger kind of similar to tanner.

### CODE REVIEW

**Comment by glaslos** [comment]
Please rebase and add a description.

**Comment by mzfr** [comment]
@glaslos I made changes :)

File: logger.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    def __init__(self, level):
+        self.level = level
+
+    def filter(self, record):
+        return record.levelno < self.level
+
+class Logger:
+    @staticmethod
+    def create_logger(debug_filename, err_filename, logger_name):
+
```
Comment (glaslos) [inline logger.py]: Empty space?

File: logger.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+        logger = logging.getLogger(logger_name)
+        logger.setLevel(logging.DEBUG)
```
Comment (glaslos) [inline logger.py]: Please expose the log level

File: logger.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        logger.propagate = False
+        formatter = logging.Formatter(fmt='%(asctime)s %(levelname)s:%(name)s:%(funcName)s: %(message)s',
+                                      datefmt='%Y-%m-%d %H:%M')
```
Comment (glaslos) [inline logger.py]: Add seconds

File: logger.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+        # ERROR log to 'snare.err'
+        error_log_handler = logging.handlers.RotatingFileHandler(err_filename, encoding='utf-8')
+        error_log_handler.setLevel(logging.ERROR)
+        error_log_handler.setFormatter(formatter)
+        logger.addHandler(error_log_handler)
+
+        # DEBUG log to 'snare.log'
+        debug_log_handler = logging.handlers.RotatingFileHandler(debug_filename, encoding='utf-8')
+        debug_log_handler.setLevel(logging.DEBUG)
```
Comment (glaslos) [inline logger.py]: You first set the level to DEBUG and then you filter out everything below ERROR?

File: logger.py [hunk 5]
```diff
@@ -10,27 +10,21 @@ def filter(self, record):
 
 class Logger:
     @staticmethod
-    def create_logger(debug_filename, err_filename, logger_name):
-
-
+    def create_logger(log_filename, logger_name, level_name):
         logger = logging.getLogger(logger_name)
-        logger.setLevel(logging.DEBUG)
+        logging_levels = {'CRITICAL': 50, 'ERROR': 40, 'WARNING': 30, 'INFO': 20, 'DEBUG': 10, 'NOTSET': 0}
```
Comment (glaslos) [inline logger.py]: you don't need to pass integers, the name of the level is fine.

File: logger.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+class Logger:
+    @staticmethod
+    def create_logger(log_filename, logger_name, level_name):
+        logger = logging.getLogger(logger_name)
+        logging_levels = {'CRITICAL': 50, 'ERROR': 40, 'WARNING': 30, 'INFO': 20, 'DEBUG': 10, 'NOTSET': 0}
+        logger.setLevel(logging_levels[level_name])
+        logger.propagate = False
+        formatter = logging.Formatter(fmt='%(asctime)s %(levelname)s:%(name)s:%(funcName)s: %(message)s',
+                                      datefmt='%Y-%m-%d %H:%M:%S')
+
```
Comment (glaslos) [inline logger.py]: two spaces

File: logger.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+        # Log to 'snare.log'
+        log_handler = logging.handlers.RotatingFileHandler(log_filename, encoding='utf-8')
+        log_handler.setLevel(logging_levels[level_name])
```
Comment (glaslos) [inline logger.py]: first you set the level here and then you overwrite it in line 26 and then you add a filter in line 27

File: snare.py [hunk 8]
```diff
@@ -445,12 +450,16 @@ def parse_timeout(timeout):
...
+    parser.add_argument("--logger", help="log of time-stamp, URL and source IP")
     args = parser.parse_args()
     base_path = '/opt/snare/'
     base_page_path = '/opt/snare/pages/'
     config = configparser.ConfigParser()
     config.read(os.path.join(base_path,args.config))
 
+    if args.logger:
+        log_file_name = "/opt/snare/snare.log"
+        logger.Logger.create_logger(log_file_name, __package__,args.logger)
```
Comment (glaslos) [inline snare.py]: missing space, read PEP8

File: snare.py [hunk 9]
```diff
@@ -445,12 +450,16 @@ def parse_timeout(timeout):
     parser.add_argument("--auto-update", help="auto update SNARE if new version available ", default=True)
     parser.add_argument("--update-timeout", help="update snare every timeout ", default='24H')
     parser.add_argument("--server-header", help="set server-header", default='nginx')
+    parser.add_argument("--logger", help="log of time-stamp, URL and source IP")
```
Comment (glaslos) [inline snare.py]: what arguments am I supposed to pass to the logger flag?

File: snare.py [hunk 10]
```diff
@@ -445,12 +450,16 @@ def parse_timeout(timeout):
...
     args = parser.parse_args()
     base_path = '/opt/snare/'
     base_page_path = '/opt/snare/pages/'
     config = configparser.ConfigParser()
     config.read(os.path.join(base_path,args.config))
 
+    if args.logger:
+        log_file_name = "/opt/snare/snare.log"
```
Comment (glaslos) [inline snare.py]: please let me use an argument to change the location.

File: logger.py [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+    def filter(self, record):
+        return record.levelno < self.level
+
+class Logger:
+    @staticmethod
+    def create_logger(log_filename, logger_name, level_name):
+        logger = logging.getLogger(logger_name)
+        logging_levels = {'CRITICAL': 50, 'ERROR': 40, 'WARNING': 30, 'INFO': 20, 'DEBUG': 10, 'NOTSET': 0}
+        logger.setLevel(logging_levels[level_name])
+        logger.propagate = False
```
Comment (glaslos) [inline logger.py]: can you explain why you set this to false?

Comment (mzfr) [reply logger.py]: I think setting it to true will attach the handler with logger and it will log same thing twice

**Comment by mzfr** [comment]
I didn't remove those integers because I think it make setting log level easy otherwise  we'll have to use something like [this](https://docs.python.org/2/library/logging.config.html) which will make logger a bit complex

**Comment by glaslos** [comment]
you think `logger.setLevel("DEBUG")` doesn't work?

File: logger.py [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+        logger = logging.getLogger(logger_name)
+        logging_levels = {'CRITICAL': 50, 'ERROR': 40,
+                          'WARNING': 30, 'INFO': 20, 'DEBUG': 10, 'NOTSET': 0}
+        logger.setLevel(logging_levels[level_name])
+        formatter = logging.Formatter(fmt='%(asctime)s %(levelname)s:%(name)s:%(funcName)s: %(message)s',
+                                      datefmt='%Y-%m-%d %H:%M:%S')
+        # Log to 'snare.log'
+        log_handler = logging.handlers.RotatingFileHandler(
+            log_filename, encoding='utf-8')
+        log_handler.setLevel(logging_levels[level_name])
```
Comment (glaslos) [inline logger.py]: why do you set the level here?

File: snare.py [hunk 13]
```diff
@@ -429,29 +442,53 @@ def parse_timeout(timeout):
...
     args = parser.parse_args()
     base_path = '/opt/snare/'
     base_page_path = '/opt/snare/pages/'
     config = configparser.ConfigParser()
-    config.read(os.path.join(base_path,args.config))
-
+    config.read(os.path.join(base_path, args.config))
 
+    if args.logger_level:
+        if args.log_file:
```
Comment (glaslos) [inline snare.py]: this is always true with the default set

Comment (mzfr) [reply snare.py]: This was fixed

**Comment by mzfr** [comment]
@glaslos Should I move this PR to develop branch?

**Comment by afeena** [comment]
@mzfr thanks for your work :)
Please look at the #109  - I merged this because it contains actual logging into the file. If you have any improvements - feel free to suggest :)

**Pull request closed without merging**